File: cpp/templates/binary_search.cpp
// # include <bits/stdc++.h>
# include <iostream>

using namespace std;

int n;
int arr[100001];
int x;

int lowerBound(vector<int> &nums, int target){
    int low = 0, high = n - 1, ans = -1;
    while (low <= high){
        int mid = ((high - low) >> 1) + low;
        // cout << mid << " " << nums[mid] << endl;
        if (nums[mid] == target)
            ans = mid;
        if (nums[mid] < target)
            low = mid + 1;
        else
            high = mid - 1;
    }
    return ans;
}

int upperBound(vector<int> &nums, int target){
    int low = 0, high = n - 1, ans = -1;
    while (low <= high){
        int mid = ((high - low) >> 1) + low;
        if (nums[mid] == target)
            ans = mid;
        if (nums[mid] > target)
            high = mid - 1;
        else
            low = mid + 1;
    }
    return ans;
}

int check(int mid) {
    if(arr[mid]>=x) return 1;
    return 0;
}

int main() {
    cin>>n;
    for(int i=0;i<n;i++)cin>>arr[i];
    cin>>x;

    // lower bound
    int lo=0;
    int hi=n-1;
    int ans=n; // default

    while(lo<=hi) {
        int mid=lo+(hi-lo)/2;
        if(check(mid)==1) {
            ans = mid;
            hi = mid-1;
        }else{
            lo = mid+1;
        }
    }
    cout << "lower bound is " << ans << endl;

    return 0;
}

File: python/templates/binary_search.py
def bs(arr,val):
    low,high=0,len(arr)-1
    while low<=high:
        mid=low+((high-low)>>1)
        if val==arr[mid]:return mid
        if val>arr[mid]:low=mid+1
        else:high=mid-1
    return -1
# res=bs(nums,target)

def bsOnAns(arr,low,high,k,ans):
    def check(val): # minimise sample
        cnt=0
        for x in arr:
            cnt+=(x+val-1)//val
            if cnt>k:return False
        return True

    # def check(val): # maximise sample
    #     rem=k
    #     cnt=0
    #     for x in arr:
    #         temp=min(rem,x)
    #         x-=temp
    #         rem-=temp
    #         cnt+=x==0
    #         if cnt>=val:return True
    #     return False

    while low<=high:
        mid=low+((high-low)>>1)
        if check(mid):
            ans=mid
            high=mid-1 # minimise
            # low=mid+1 # maximise
        else:
            low=mid+1 # minimise
            # high=mid-1 # maximise
    return ans
# bsOnAns(nums,lowerlimit,upperlimit,kvalue,defaultanswer)


# use this for getting value at a specific position in a sorted array
# usecases: kth element,median,mode,multiples,...
def bsOnAns(arr,low,high,k,ans):
        # print(f"low:{low}, high:{high}, pos:{k}")
        def check(val): # minimise sample
            cnt=0
            # print(f":{val},:{cnt}")
            return cnt>=k

        while low<=high:
            mid=low+((high-low)>>1)
            if check(mid): # valid found so update answer
                ans=mid
                high=mid-1 # look for less possible candidate
            else:
                low=mid+1
        return ans